File: lib/analysis.ts
import { Chess } from 'chess.js'
import { resolveStockfishPath, StockfishEngine } from '@/lib/stockfish'

const BLUNDER_THRESHOLD = 200
const DEFAULT_MOVE_TIME_MS = 100

export interface GameData {
  game: {
    date?: string
    white?: string
    black?: string
    result?: string
    opening_name?: string
    my_accuracy?: number
    blunders: number
    pgn_text: string
  }
  moves: Array<{
    move_number: number
    ply: number
    fen: string
    move_san: string
    engine_eval?: number
    is_blunder: boolean
  }>
}

export async function analyzePgn(
  pgnText: string,
  stockfishPath: string,
  username: string
): Promise<GameData[]> {
  if (!pgnText.trim()) {
    return []
  }

  try {
    const enginePath = resolveStockfishPath(stockfishPath)
    const moveTimeMs = parseInt(
      process.env.STOCKFISH_TIME_LIMIT_MS || '',
      10
    ) || DEFAULT_MOVE_TIME_MS
    const engine = new StockfishEngine(enginePath, moveTimeMs)
    await engine.start()

    const gameStrings = pgnText.split(/(\[Event\s+"[^"]+"\])/g).filter(s => s.trim())
    const results: GameData[] = []
    
    const games: string[] = []
    for (let i = 0; i < gameStrings.length; i++) {
      if (gameStrings[i].startsWith('[Event')) {
        games.push(gameStrings[i] + (gameStrings[i+1] || ''))
        i++
      } else {
        games.push(gameStrings[i])
      }
    }

    try {
      for (const gamePgn of games) {
        const chess = new Chess()
        try {
          chess.loadPgn(gamePgn)
        } catch (e) {
          console.warn('Failed to load PGN:', e)
          continue
        }

        const history = chess.history({ verbose: true })
        const moves: GameData['moves'] = []
        const losses: number[] = []
        let blunders = 0
        
        const tempChess = new Chess()
        let ply = 0

        const headers = chess.header()
        const normalizedUsername = username.trim().toLowerCase()
        const isWhite = headers.White?.toLowerCase() === normalizedUsername
        const isBlack = headers.Black?.toLowerCase() === normalizedUsername
        const userColor = isWhite ? 'white' : isBlack ? 'black' : null

        for (const move of history) {
          try {
            const isUserMove = userColor
              ? (userColor === 'white' && ply % 2 === 0) ||
                (userColor === 'black' && ply % 2 === 1)
              : false

            let engineEval: number | undefined
            let isBlunder = false

            if (isUserMove) {
              const evalBefore = await engine.evaluate(
                tempChess.fen(),
                tempChess.turn()
              )
              tempChess.move(move.san)
              const evalAfter = await engine.evaluate(
                tempChess.fen(),
                tempChess.turn()
              )
              engineEval = evalAfter
              const loss = centipawnLoss(evalBefore, evalAfter, userColor)
              losses.push(loss)
              if (loss > BLUNDER_THRESHOLD) {
                isBlunder = true
                blunders++
              }
            } else {
              tempChess.move(move.san)
            }

            ply++
            const moveNumber = Math.ceil(ply / 2)

            moves.push({
              move_number: moveNumber,
              ply,
              fen: tempChess.fen(),
              move_san: move.san,
              engine_eval: engineEval,
              is_blunder: isBlunder,
            })
          } catch (e) {
            console.warn('Failed move in game:', e)
          }
        }

        const accuracy = accuracyFromLosses(losses)

        results.push({
          game: {
            date: headers.Date ?? undefined,
            white: headers.White ?? undefined,
            black: headers.Black ?? undefined,
            result: headers.Result ?? undefined,
            opening_name: headers.Opening ?? undefined,
            my_accuracy: accuracy,
            blunders,
            pgn_text: gamePgn,
          },
          moves,
        })
      }
    } finally {
      await engine.stop()
    }

    return results
  } catch (error) {
    console.error('Error analyzing PGN:', error)
    throw new Error('Failed to analyze PGN: ' + (error as Error).message)
  }
}

export async function parsePgnWithoutEngine(pgnText: string): Promise<GameData[]> {
  if (!pgnText.trim()) {
    return []
  }

  const gameStrings = pgnText.split(/(\[Event\s+"[^"]+"\])/g).filter(s => s.trim())
  const results: GameData[] = []

  const games: string[] = []
  for (let i = 0; i < gameStrings.length; i++) {
    if (gameStrings[i].startsWith('[Event')) {
      games.push(gameStrings[i] + (gameStrings[i + 1] || ''))
      i++
    } else {
      games.push(gameStrings[i])
    }
  }

  for (const gamePgn of games) {
    const chess = new Chess()
    try {
      chess.loadPgn(gamePgn)
    } catch (e) {
      console.warn('Failed to load PGN:', e)
      continue
    }

    const history = chess.history({ verbose: true })
    const moves: GameData['moves'] = []
    const tempChess = new Chess()
    let ply = 0

    for (const move of history) {
      try {
        tempChess.move(move.san)
        ply++
        const moveNumber = Math.ceil(ply / 2)
        moves.push({
          move_number: moveNumber,
          ply,
          fen: tempChess.fen(),
          move_san: move.san,
          engine_eval: undefined,
          is_blunder: false,
        })
      } catch (e) {
        console.warn('Failed move in game:', e)
      }
    }

    const headers = chess.header()

    results.push({
      game: {
        date: headers.Date ?? undefined,
        white: headers.White ?? undefined,
        black: headers.Black ?? undefined,
        result: headers.Result ?? undefined,
        opening_name: headers.Opening ?? undefined,
        my_accuracy: undefined,
        // IMPORTANT: We have no engine evaluation here, so blunders are UNKNOWN.
        // Use a sentinel value to avoid lying with "0".
        blunders: -1,
        pgn_text: gamePgn,
      },
      moves,
    })
  }

  return results
}

function centipawnLoss(
  beforeCp: number,
  afterCp: number,
  myColor: 'white' | 'black' | null
): number {
  if (!myColor) return 0
  const beforePov = myColor === 'white' ? beforeCp : -beforeCp
  const afterPov = myColor === 'white' ? afterCp : -afterCp
  const loss = beforePov - afterPov
  return Math.max(0, Math.round(loss))
}

function accuracyFromLosses(losses: number[]): number | undefined {
  if (losses.length === 0) return undefined
  const avgLoss = losses.reduce((sum, loss) => sum + loss, 0) / losses.length
  const accuracy = 100 - avgLoss / 2
  return Math.max(0, Math.min(100, accuracy))
}

File: src/analysis.py
import io
import os
import shutil
from typing import Dict, List, Optional

import chess
import chess.engine
import chess.pgn

from urllib.parse import unquote, urlparse

ENGINE_TIME_LIMIT = 0.1
BLUNDER_THRESHOLD = 200
MATE_SCORE = 100000


def resolve_stockfish_path(stockfish_path: str) -> str:
    candidates = [stockfish_path, "./stockfish", "stockfish"]
    for candidate in candidates:
        if not candidate:
            continue
        if os.path.isfile(candidate):
            return candidate
        resolved = shutil.which(candidate)
        if resolved:
            return resolved
    raise FileNotFoundError(
        "Stockfish binary not found. Provide a valid path in the sidebar."
    )


def identify_user_color(headers: Dict[str, str], username: str) -> Optional[str]:
    if not username:
        return None
    normalized = username.strip().lower()
    if normalized and headers.get("White", "").strip().lower() == normalized:
        return "white"
    if normalized and headers.get("Black", "").strip().lower() == normalized:
        return "black"
    return None


def score_to_cp(score: chess.engine.PovScore) -> int:
    return int(score.score(mate_score=MATE_SCORE))


def evaluate_position(board: chess.Board, engine: chess.engine.SimpleEngine) -> int:
    info = engine.analyse(board, chess.engine.Limit(time=ENGINE_TIME_LIMIT))
    return score_to_cp(info["score"].pov(chess.WHITE))


def centipawn_loss(before_cp: int, after_cp: int, my_color: str) -> int:
    before_pov = before_cp if my_color == "white" else -before_cp
    after_pov = after_cp if my_color == "white" else -after_cp
    loss = before_pov - after_pov
    return int(max(0, loss))


def accuracy_from_losses(losses: List[int]) -> Optional[float]:
    if not losses:
        return None
    avg_loss = sum(losses) / len(losses)
    accuracy = 100.0 - (avg_loss / 2.0)
    return float(max(0.0, min(100.0, accuracy)))


def analyze_game(
    game: chess.pgn.Game, engine: chess.engine.SimpleEngine, username: str
) -> Dict:
    headers = game.headers
    my_color = identify_user_color(headers, username)

    board = game.board()
    moves: List[Dict] = []
    losses: List[int] = []
    blunders = 0
    ply = 0

    for move in game.mainline_moves():
        ply += 1
        move_number = board.fullmove_number
        move_san = board.san(move)
        is_user_move = (
            my_color == "white"
            and board.turn == chess.WHITE
            or my_color == "black"
            and board.turn == chess.BLACK
        )
        engine_eval = None
        is_blunder = False

        if is_user_move:
            eval_before = evaluate_position(board, engine)
            board.push(move)
            eval_after = evaluate_position(board, engine)
            engine_eval = eval_after
            loss = centipawn_loss(eval_before, eval_after, my_color)
            losses.append(loss)
            if loss > BLUNDER_THRESHOLD:
                is_blunder = True
                blunders += 1
        else:
            board.push(move)

        moves.append(
            {
                "move_number": move_number,
                "ply": ply,
                "fen": board.fen(),
                "move_san": move_san,
                "engine_eval": engine_eval,
                "is_blunder": is_blunder,
            }
        )

    game_pgn = game.accept(
        chess.pgn.StringExporter(headers=True, variations=False, comments=False)
    )

    return {
        "game": {
            "date": headers.get("Date"),
            "white": headers.get("White"),
            "black": headers.get("Black"),
            "result": headers.get("Result"),
            "opening_name": derive_opening_name(headers),
            "my_accuracy": accuracy_from_losses(losses),
            "blunders": blunders,
            "pgn_text": game_pgn,
        },
        "moves": moves,
    }


def analyze_pgn(
    pgn_text: str, stockfish_path: str, username: str
) -> List[Dict]:
    if not pgn_text.strip():
        return []

    engine_path = resolve_stockfish_path(stockfish_path)
    results: List[Dict] = []

    with chess.engine.SimpleEngine.popen_uci(engine_path) as engine:
        stream = io.StringIO(pgn_text)
        while True:
            game = chess.pgn.read_game(stream)
            if game is None:
                break
            results.append(analyze_game(game, engine, username))

    return results


def derive_opening_name(headers: Dict[str, str]) -> Optional[str]:
    opening = normalize_opening_name(headers.get("Opening"))
    if opening:
        return opening

    eco_url = normalize_opening_name(headers.get("ECOUrl"))
    if eco_url:
        eco_url_name = opening_name_from_eco_url(eco_url)
        if eco_url_name:
            return eco_url_name

    eco = normalize_opening_name(headers.get("ECO"))
    if eco:
        return f"ECO {eco}"

    return None


def normalize_opening_name(value: Optional[str]) -> Optional[str]:
    if not value:
        return None
    trimmed = value.strip()
    return trimmed if trimmed else None


def opening_name_from_eco_url(eco_url: str) -> Optional[str]:
    try:
        parsed = urlparse(eco_url)
        path = parsed.path.strip("/")
        if not path:
            return None
        last_segment = path.split("/")[-1]
        if not last_segment:
            return None
        decoded = unquote(last_segment)
        normalized = decoded.replace("-", " ").replace("_", " ").strip()
        normalized = " ".join(normalized.split())
        return normalized if normalized else None
    except Exception:
        return None
